src/devices/src/virtio/console/inorder_handler.rs

```diff
@@ -0,0 +1,97 @@
+use std::io::Write;
+use std::result;
+use virtio_queue::{Queue};
+use vm_memory::{GuestAddressSpace};
+use crate::virtio::{SignalUsedQueue};
+use virtio_console::console;
+
+#[derive(Debug)]
+pub enum Error {
+    GuestMemory(vm_memory::GuestMemoryError),
+    Queue(virtio_queue::Error),
+}
+
+impl From<vm_memory::GuestMemoryError> for Error {
+    fn from(e: vm_memory::GuestMemoryError) -> Self {
+        Error::GuestMemory(e)
+    }
+}
+
+impl From<virtio_queue::Error> for Error {
+    fn from(e: virtio_queue::Error) -> Self {
+        Error::Queue(e)
+    }
+}
+
+
+pub struct InOrderQueueHandler<M: GuestAddressSpace, S: SignalUsedQueue, T: Write> {
+    pub driver_notify: S,
+    pub transmitq: Queue<M>,
+    pub receiveq: Queue<M>,
+    pub output: T,
+    pub console: console::Console,
+}
+
+impl<M, S, T> InOrderQueueHandler<M, S, T>
+where
+    M: GuestAddressSpace,
+    S: SignalUsedQueue,
+    T: Write,
+{
+    pub fn process_transmitq(&mut self) -> result::Result<(), Error>{
+        // To see why this is done in a loop, please look at the `Queue::enable_notification`
+        // comments in `virtio_queue`.
+        loop {
+            self.transmitq.disable_notification()?;
+
+            while let Some(mut chain) = self.transmitq.iter()?.next() {
+                let read_len = self.console.process_transmitq_chain(&mut chain, &mut self.output);
+
+                self.transmitq.add_used(chain.head_index(), read_len as u32)?;
+            }
+            if !self.transmitq.enable_notification()? {
+                break;
+            }
+        }
+        if self.transmitq.needs_notification()? {
+            self.driver_notify.signal_used_queue(1);
+        }
+
+        Ok(())
+    }
+
+    pub fn process_receiveq(&mut self) -> result::Result<(), Error>{
+        // To see why this is done in a loop, please look at the `Queue::enable_notification`
+        // comments in `virtio_queue`.
+
+        let mut notify = false;
+
+        loop {
+            if self.console.is_input_buffer_empty() {
+                break;
+            }
+            self.receiveq.disable_notification()?;
+
+            while let Some(mut chain) = self.receiveq.iter()?.next() {
+                let used_len = self.console.process_receiveq_chain(&mut chain);
```

Comment:
In case we weren't able to write any byte to the memory buffer, we might want to go back one position into the queue. I think this is how CH and crosvm are handling this situation.